File: schemas/analytics-metrics/drizzle/saved_reports.ts
// saved_reports: User-created report configurations with visibility controls.
// See README.md for full design rationale.

import {
  pgTable,
  pgEnum,
  uuid,
  text,
  jsonb,
  timestamp,
  index,
} from "drizzle-orm/pg-core";
import { users } from "../../auth-rbac/drizzle/users";

export const visibilityEnum = pgEnum("visibility", [
  "private",
  "team",
  "public",
]);

export const savedReports = pgTable(
  "saved_reports",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    name: text("name").notNull(),
    description: text("description"),
    config: jsonb("config").notNull(),
    visibility: visibilityEnum("visibility").notNull().default("private"),
    createdBy: uuid("created_by").notNull().references(() => users.id, { onDelete: "cascade" }),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true }).notNull().defaultNow().$onUpdateFn(() => new Date()),
  },
  (table) => [
    index("idx_saved_reports_created_by").on(table.createdBy),
    index("idx_saved_reports_visibility").on(table.visibility),
  ]
);

File: schemas/analytics-metrics/spacetimedb/saved_reports.rs
// saved_reports: Persisted report configurations with visibility controls.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum Visibility {
    Private, // type: String
    Team,
    Public,
}

#[spacetimedb::table(name = saved_reports, public)]
pub struct SavedReport {
    #[primary_key]
    pub id: String, // UUID

    pub name: String,
    pub description: Option<String>,
    pub config: String, // JSON
    #[index(btree)]
    pub visibility: Visibility,

    #[index(btree)]
    pub created_by: String, // UUID, FK → users.id (cascade delete)

    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
